source/language/classical.rst

```diff
@@ -184,26 +189,85 @@ may optionally be followed by ``else { program }``.
-to ``name``. The for loop body is not permitted to modify the loop variable of
-the indexset.
+For loops
+.........
+
+The statement ``for <type> <name> in <values> <body>`` loops over the
+items in ``values``, assigning each value to the variable ``name`` in subsequent
+iterations of the loop ``body``.  ``values`` can be:
+
+- a discrete set of scalar types, defined using the
+  :ref:`array-literal syntax <types-arrays>`, such as ``{1, 2, 3}``.  Each value
+  in the set must be able to be implicitly promoted to the type ``type``.
+
+- a range expression in square brackets of the form ``[start : (step :)? stop]``,
+  where ``step`` is equal to ``1`` if omitted.  As in other range expressions,
+  the range is inclusive at both ends.  Both ``start`` and ``stop`` must be
+  given.  All three values must be of integer or unsigned-integer types.  The
+  scalar type of elements in the resulting range expression is the same as the
+  type of result of the :ref:`implicit promotion <implicit-promotion-rules>`
+  between ``start`` and ``stop``.  For example, if ``start`` is a ``uint[8]``
+  and ``stop`` is an ``int[16]``, the values to be assigned will all be of type
+  ``int[16]``.
+
+- a value of type ``bit[n]``, or the target of a ``let`` statement that creates
+  an alias to classical bits.  The corresponding scalar type of the loop
+  variable is ``bit``, as appropriate.
+
+- a value of type ``array[<scalar>, n]``, _i.e._ a one-dimensional
+  array.  Values of type ``scalar`` must be able to be implicitly promoted to
+  values of type ``type``.  Modification of the loop variable does not change
+  the corresponding value in the array.
+
+The loop body can either be a single statement terminated by a semicolon, or a
+program block in curly braces ``{}`` containing several statements.
+
+Assigning a value to the loop variable within an iteration over the body does
+not affect the next value that the loop variable will take.
+
+The scope of the loop variable is limited to the body of the loop.  It is not
+accessible after the loop.
 
 .. code-block:: c
 
-   int[32] b;
+   int[32] b = 0;
    // loop over a discrete set of values
-   for i in {1, 5, 10} {
+   for int[32] i in {1, 5, 10} {
        b += i;
-   } // b == 16
+   }
+   // b == 16, and i is not in scope.
+
+   // loop over every even integer from 0 to 20 using a range, and call a
+   // subroutine with that value.
+   for int i in [0:2:20]
+      subroutine(i);
 
-   // loop over every even integer from 0 to 20 using an indexset
-   for i in [0:2:20] {
+   // high precision typed loop variable
+   for uint[64] i in [4294967296:4294967306] {
       // do something
    }
 
-The statement ``while ( bool ) { program }`` executes program until the Boolean evaluates to
+   // Loop over an array of floats.
+   array[float[64], 4] my_floats = {1.2, -3.4, 0.5, 9.8};
+   for float[64] f in my_floats {
+      // do something with 'f'
+   }
+
+   // Loop over a register of bits.
```

Comment:
Is there a guaranteed order in which the iteration hits the bits? Relevant in the situation that the body has side effects

Reply:
I had assumed that registers have an order since they can be indexed with successive integers, and it'd be in that order.  I can make that explicit, assuming general agreement - it was an oversight that I didn't mention it.

Reply:
That example was meant to assert that names bound in `let` statements are valid targets for iteration, rather than forbidding slicing.  I think that the question of slicing is more just one of the type system in general - if the type of `a[b:c]` is the same as something that can be iterated over, then it should be.  Right now that gets into funny questions about what the type of that expression is for registers - the existence of `let` makes things pretty weird.